service/history/workflow/mutable_state_impl.go

```diff
@@ -2910,36 +2985,47 @@ func (ms *MutableStateImpl) addBuildIdToLoadedSearchAttribute(
 	stamp *commonpb.WorkerVersionStamp,
 ) []string {
 	var newValues []string
+	var buildId string
 	effectiveBehavior := ms.GetEffectiveVersioningBehavior()
+
+	// set up the unversioned or assigned:x sentinels (versioning v2)
 	if !stamp.GetUseVersioning() && effectiveBehavior == enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED { // unversioned workflows may still have non-nil deployment, so we don't check deployment
 		newValues = append(newValues, worker_versioning.UnversionedSearchAttribute)
-	} else if effectiveBehavior == enumspb.VERSIONING_BEHAVIOR_PINNED {
-		newValues = append(newValues, worker_versioning.PinnedBuildIdSearchAttribute(ms.getPinnedDeployment()))
 	} else if ms.GetAssignedBuildId() != "" {
 		newValues = append(newValues, worker_versioning.AssignedBuildIdSearchAttribute(ms.GetAssignedBuildId()))
 	}
 
+	// get the most up-to-date pinned entry put it at the front (v3 reachability and v3.1 drainage)
+	if effectiveBehavior == enumspb.VERSIONING_BEHAVIOR_PINNED {
+		if ms.GetExecutionInfo().GetVersioningInfo().GetVersion() == "" { // deployments
+			newValues = append(newValues, worker_versioning.PinnedBuildIdSearchAttribute(ms.getPinnedDeployment(), ""))
+		} else { // worker deployments
+			newValues = append(newValues, worker_versioning.PinnedBuildIdSearchAttribute(nil, ms.getPinnedVersion()))
+		}
+	}
+
+	// get the build id entry (all versions of versioning)
 	if effectiveBehavior == enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED {
-		buildId := worker_versioning.VersionStampToBuildIdSearchAttribute(stamp)
-		found := slices.Contains(newValues, buildId)
-		for _, existingValue := range existingValues {
-			if existingValue == buildId {
-				found = true
-			}
-			if !worker_versioning.IsUnversionedOrAssignedBuildIdSearchAttribute(existingValue) {
-				newValues = append(newValues, existingValue)
-			}
+		buildId = worker_versioning.VersionStampToBuildIdSearchAttribute(stamp)
+	} else { // versioning behaviors
+		buildId = worker_versioning.VersionedBuildIdSearchAttribute(ms.getVersionedBuildId())
+	}
```

Comment:
I think you should only add build ID if the stamp is given. no need to extract build ID using VersionedBuildIdSearchAttribute.
Note that `versioned:..` `unversioned:..` are supposed to be populated only for builds which completed wfts.

Instead of this you should ensure that afterAddWorkflowTaskCompletedEvent calls updateBuildIdsAndDeploymentSearchAttributes with a stamp (which it does). But what we need to additionally cover is to make the updateBuildIdsAndDeploymentSearchAttributes call with a stamp made out of attrs.VersioningBehavior and attrs.WorkerDeploymentVersion even if attrs.GetWorkerVersion() is nill. (Note that we've deprecated WorkerVersionStamp since v3.1)
Once you do ^ then you can only worry about adding `versioned:..` `unversioned:..` items IF stamp is not nil.

Reply:
ah ok. makes sense, let me check